widget/numerical_entry.go

```diff
@@ -72,12 +177,253 @@ func (e *NumericalEntry) Keyboard() mobile.KeyboardType {
+			return r, true
+		}
+	case ' ': // space
+		fallthrough
+	case 0xa0: // non-breaking space
+		if e.thouSep == ' ' || e.thouSep == 0xa0 {
+			return e.thouSep, true
+		}
+	case '\'': // single quote
+		fallthrough
+	case 0x2019: // right single quote mark
+		if e.thouSep == '\'' || e.thouSep == 0x2019 {
+			return e.thouSep, true
+		}
+	default:
+		if r == e.radixSep || r == e.thouSep {
+			return r, true
+		}
+	}
+	return 0, false
+}
+
+// getValidText filters the input text to allow only valid characters
+// based on the current locale and the entry's configuration.
+func (e *NumericalEntry) getValidText(curText, text string) string {
+	var s strings.Builder
+	for i, r := range text {
+		rn, ok := e.getRuneForLocale(r)
+		if !ok {
+			continue
+		}
+		if rn == e.minus {
+			if !e.AllowNegative || !(len(curText) == 0) || i != 0 {
+				continue
+			}
+		}
+		if rn == e.radixSep {
+			if !e.AllowFloat {
+				continue
+
+			}
+
+		}
+		s.WriteRune(rn)
+	}
+	return s.String()
+}
+
+// makeParsable prepares the text for parsing by removing thousand separators,
+// replacing the minus sign with a standard minus, and replacing the radix
+// separator with a period. It also validates the text before processing.
+func (e *NumericalEntry) makeParsable(text string) (string, error) {
+	err := e.ValidateText(text)
+	if err != nil {
+		return "", err
+	}
+	t := text
+	t = strings.ReplaceAll(text, string(e.thouSep), "")
+	t = strings.Replace(t, string(e.minus), "-", -1)
+	t = strings.Replace(t, string(e.radixSep), ".", -1)
+	return t, nil
+}
+
+// getLocaleRunes determines the minus sign, radix, and thousand separator
+// characters for a given locale by formatting a number and extracting
+// the relevant characters. It returns the minus sign, radix, and thousand
+// separator runes.
+func (e *NumericalEntry) getLocaleRunes(locale string) {
+	e.minus = '-'
+	e.thouSep = ','
+	e.radixSep = '.'
+	lang, err := language.Parse(locale)
+	if err != nil {
+		fyne.LogError("Language parse error: ", err)
+		return
+	}
+	p := message.NewPrinter(lang)
+	numStr := p.Sprintf("%f", -12345.5678901)
+	runes := []rune(numStr)
+
+	// Define helper function to find separator
+	findSeparator := func(runes []rune) rune {
+		for _, r := range runes {
+			if !unicode.IsDigit(r) && r != e.minus {
+				return r
+			}
+		}
+		return 0 // Return 0 if no separator is found
+	}
+
+	// first rune is the "minus" sign
+	if len(runes) > 0 {
+		e.minus = runes[0]
+	}
+
+	// look for thousands separator
+	if len(runes) > 1 {
+		e.thouSep = findSeparator(runes[1:min(6, len(runes))])
+	}
+
+	// look for radix separator
+	if len(runes) > 6 {
+		e.radixSep = findSeparator(runes[6:])
+		if e.radixSep == 0 && e.thouSep != 0 {
+			// If no radix separator is found, and thousand separator is present,
+			// use the last non-digit character as radix separator
+			e.radixSep = findSeparator(runes[len(runes)-7:])
+		}
+	}
+}
+
+// min returns the smaller of two integers
```

Comment:
Please put utility functions at the end of the file

Reply:
Moved to end of file.